2540 filter complex queries

entantoencuanto opened the pull request:
Closes #2540


## :v: What does this PR do?

* Allows the use of a notation in the query params of API index request including operators such as `lt`, `lteq`, `gt`, `gteq`, `in` and `like`
* Parses operators provided in the request to build the appropriate query in the `GobiertoCommon::CustomFieldQuery` `filter` action.

## :mag: How should this be manually tested?

Visit API index project path. The old notation of filters in the query string with the form: `/?filter[uid]=value&filter[uid2]=1000` or adding to the body the JSON:

```json
{
    "filter": {
        "uid": "value",
        "uid2": 1000
    }
}
```
assumes that the operator is `eq`. To use other operators, as `lt`, use a LHS brackets notation:
`/?filter[uid][lt]=value&filter[uid][gt]=value2&filter[uid2][gteq]=1000`, or in JSON format:

```json
{
    "filter": {
        "uid": {
            "lt": "value",
            "gt": "value2"
        },
        "uid2": {
            "gteq": 1000
        }
    }
}
```
The `in` operator expects a string formed by the concatenation of values separated by commas.

## :shipit: Does this PR changes any configuration file?

No

(Changes in these files might need to update the role in Ansible)

## :book: Does this PR require updating the documentation?

No

entantoencuanto committed:
Ensure the format of payload and reject invalid ids in vocabulary_options custom fields

entantoencuanto committed:
Extend tests for vocabulary_options custom field

entantoencuanto committed:
Add operators to custom fields queries and API filters

entantoencuanto committed:
Add some tests and fixtures for projects filters

ferblape commented:
1. Could you give an example of how to filter projects within a range of "fecha-inicio" and "fecha-fin"?

I've tested http://talavan.gobify.net/gobierto_investments/api/v1/projects?filter[fecha-inicio][lte]=2019-08-01 but the query returns nulls too

2. Could you give an example of [in] operation in the URL?

entantoencuanto committed:
Fix filter queries with in operator

entantoencuanto commented:
> 1. Could you give an example of how to filter projects within a range of "fecha-inicio" and "fecha-fin"?

For example: http://talavan.gobify.net/gobierto_investments/api/v1/projects?filter[fecha-inicio][gteq]=2019-09-01&filter[fecha-inicio][lt]=2019-10-01

> I've tested [http://talavan.gobify.net/gobierto_investments/api/v1/projects?filter[fecha-inicio][lte]=2019-08-01](http://talavan.gobify.net/gobierto_investments/api/v1/projects?filter%5Bfecha-inicio%5D%5Blte%5D=2019-08-01) but the query returns nulls too

The correct name operator is `lteq`. Name operators different from lt, lteq, gt, gteq, in or like are ignored

> 
> 1. Could you give an example of [in] operation in the URL?

Yes! There was a problem with some changes I introduced at last moment. Fixed now: http://talavan.gobify.net/gobierto_investments/api/v1/projects?filter[presupuesto][in]=0,1111

entantoencuanto committed:
Add filters for eq, in and like operators

ferblape commented:
Can we merge this?

entantoencuanto commented:
Yes, its ready
